designcompose/src/main/java/com/android/designcompose/Layout.kt

```diff
@@ -575,9 +591,23 @@ internal fun designMeasurePolicy(view: View, layoutId: Int) =
             Log.d(TAG, "designMeasurePolicy error: null layout $name layoutId $layoutId")
         // Get width and height from constraints if they are fixed. Otherwise get them from layout.
         val myWidth =
-            if (constraints.hasFixedWidth) constraints.maxWidth else myLayout?.width() ?: 0
+            if (constraints.hasFixedWidth && constraints.maxWidth != 0) constraints.maxWidth
+            else myLayout?.width() ?: 0
         val myHeight =
-            if (constraints.hasFixedHeight) constraints.maxHeight else myLayout?.height() ?: 0
+            if (constraints.hasFixedHeight && constraints.maxHeight != 0) constraints.maxHeight
+            else myLayout?.height() ?: 0
+
+        // If constraints have forced a size that does not match our layout size, set this size into
+        // the layout manager so that it can set this size in the Rust layout manager and then
+        // recalculate any layouts that many have changed. However, only do this if the layout state
+        // has not changed after computing child layouts. This is necessary in the case where
+        // calling setNodeSize on a node causes an ancestor to resize.
+        val layoutState2 = LayoutManager.getLayoutState(layoutId)
+        if (
+            layoutState == layoutState2 &&
+                (myWidth != myLayout?.width() || myHeight != myLayout?.height())
+        )
+            LayoutManager.setNodeSize(layoutId, rootLayoutId, myWidth, myHeight)
```

Comment:
Does this end up causing a second layout? Or is this only applied in rendering?

Reply:
I guess I mean: does this mean we have to run another iteration of Rust layout? I think we're consulting the Rust-computed layout values above, and then we're calling `setNodeSize`, which I think doesn't take effect until the next iteration of layout.

I'm trying to figure out if there's any recursive stuff here, which might result in more than one additional layout iteration being needed (like Rust layout generates a value, it is used to calculate incoming Constraints because a parent uses the value for child placement, the child then uses it via incoming Constraints... it looks like the child is always placed using a value directly from Rust layout, so I think that means if one layout iteration causes a node size to change, then the children might get a new size on the next iteration, which could cause a further layout iteration if one of those children has a layout modifier.

I guess the layout modifier needs to know the computed size -- so we can't create a pretend size and run it through the Modifier up front. I'm wondering if there's any way to make this more efficient or get it down to a single pass, or if we have to define our own Modifiers that are analogues to the Compose ones (which would be a pity from an API perspective).

Reply:
I'm still testing this which is why it's in a draft state, but I think it probably does cause a second layout pass. If say, a call to `fillMaxSize()` causes `#stage` to resize from 200x200 to 400x400, it Rust layout will compute it first as 200x200. Then the layout function will see that constraints have changed it to 400x400, so it calls `setNodeSize(400, 400)`. This calls into Rust layout and recomputes layout for children, then sends the new layouts back. However by this point, the layout for `#stage` has already measured all children with the old layouts. The new layouts with a new `layoutState` ID will then cause all nodes that changed to call layout again.

Reply:
We discussed online recently and thought it might be possible to use the measure func to glue the measurement of a custom composable (or customized child) into the rust layout world in a single pass, to avoid the complexity and performance issue with a multi-pass solution. I think there's also nothing preventing us from rendering a frame that's mid-way through layout, here, too, which isn't ideal.

Reply:
As we discussed offline, this is not possible with the way the code is currently structured because if a child changes, there's no way for it to force the root node to run layout again.